fix: erroneous "new" minion presence events on process startup

raddessi opened the pull request:
### What does this PR do?
This fixes the issue of the master processes sending incorrect presence events on either the restart of the salt-master process or any of the `Maintenance` process restart events (default is every hour). This issue began in 3006.0.

Although this was intended to fix the hourly incorrect presence update this also fixes the spammy presence updates between  salt-master process restarts. 

### What issues does this PR fix or reference?
Fixes: #64505

### Previous Behavior
The salt-master process sends incorrect presence events at every `restart_interval` of the `Maintenance` process, signalling that every node connected to the master is "new".

### New Behavior
Only actual presence events are sent.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
  - I don't see any references to the `Maintenance` process in the current documentation so I think this does not need a specific mention since we are fixing a bug
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

I will update all of these once this is commented on and direction is approved.

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

raddessi commented:
Just as background info, I noticed this because I use the slack engine and presence reactor.

whytewolf commented:
so, this should not be labeled a feature, or it should be going into master. if this is a fix for a bug it should be labeled as such. if it is a new feature it needs to go into master. 

Next, it is going to need a set of tests and changelog. as well as fixing the pre-commit which just looks like an isort issue.

raddessi commented:
Good point. I started writing this as a feature but it does appear to be more of a bug fix at this point. I'll swap the title thank you. 

I'll add tests etc once I get approval on the concept. I don't want to spend time on that if this gets denied

Ch3LL commented:
@dwoz can you give any input on @raddessi 's current approach before he adds tests?

raddessi commented:
@dwoz bumping this for design review, thank you

raddessi commented:
> @dwoz bumping this for design review, thank you

bumping again @dwoz

dwoz commented:
- Should we use a JSON file at all?

I'm okay with it, but we should look at what format other cache files (such as pillar cache) use. I suspect pillar cache is using msgpack.

- If so, should it contain only this list and be named something like the current presence-data or should it be more generic for all of the maintenance process and be named something like maintenance-data and contain a dict that the presence data is stored within?

I'm fine with the presence-data name.

- If we do use a file, where is the correct place for it to be kept?

Seems reasonable but how do minions get removed from the file? What happens if a minion goes offline when the master is offline and can't receive an event? Is there something like a ttl on the minions in that file?

raddessi commented on salt/master.py:
```diff
@@ -258,7 +258,16 @@ def run(self):
+        # load the presence data from the cache on disk
+        presence_cache = salt.utils.cache.CacheFactory.factory(
+            "disk",
+            3600,
```
the value of 3600 doesn't really matter all that much.. theoretically we could increase it to handle longer shutdowns of the salt master but that's probably an edge case, and from my testing I think there is  bug in the current `CacheDisk` class anyway where I think the `CacheDisk_cachetime` key is being stored as the human readable but completely valueless `present` rather than an actual timestamp. I'd rather not have to fix that myself here.

raddessi replied:
Ah ok there is no bug. I just was seeign the key rather than the value. All good here.

raddessi commented:
> * Should we use a JSON file at all?
> 
> I'm okay with it, but we should look at what format other cache files (such as pillar cache) use. I suspect pillar cache is using msgpack.

OK I changed this over to use the same storage that I think the pillar cache does, `salt.utils.cache.CacheFactory`


> * If we do use a file, where is the correct place for it to be kept?
> 
> Seems reasonable but how do minions get removed from the file? What happens if a minion goes offline when the master is offline and can't receive an event? Is there something like a ttl on the minions in that file?

The entire contents of the cache gets replaced with whatever the current presence list is so the only time a ttl comes in to play is if the master gets shut off for a period of time. I don't think we would want any other behavior because it could mask underlying issues with minions flapping.

raddessi commented:
Sorry for the delay in updates @dwoz byt the time I had pinged you I had almost given up on this and have been away from github for a while. I'll start checking back in again. Thanks!

raddessi commented:
Apologies for the bump, can I get another review when you have time @dwoz - Thanks!

raddessi commented:
ah my logic is off, I will fix

raddessi commented:
bumping this @dwoz @cmcmarrow

raddessi commented:
test added

raddessi commented:
Changelog entry added, I don't think we need docs on this one as I said above. Ready to go if you think this is the right approach :crossed_fingers: 

Not that I would like to add more work for myself but does this need to be able to wipe the cache file if it becomes corrupted for some reason?

raddessi commented on tests/pytests/unit/test_master.py:
```diff
@@ -1,10 +1,29 @@
 
 import salt.master
 import salt.utils.platform
-from tests.support.mock import patch
+from tests.support.mock import MagicMock, patch
+
+
+@pytest.fixture
+def maintenance_opts(master_opts):
+    """
+    Options needed for master's Maintenence class
+    """
+    opts = master_opts.copy()
+    opts.update(git_pillar_update_interval=180, maintenance_interval=181)
+    yield opts
+
+
+@pytest.fixture
+def maintenance(maintenance_opts):
+    """
+    The master's Maintenence class
+    """
+    return salt.master.Maintenance(maintenance_opts)
 
```
I copied this in from `master`

raddessi commented:
wait this needs to check for the change event also dammit. Will fix

dwoz commented:
Are these conflicts able to be resolved?

dwoz committed:
Fix erroneous "new" minion presence events on Maintenance process restart

Persist the master Maintenance loop's connected-minion set to a disk
cache so that when the Maintenance process restarts (default: every
hour) it can seed old_present from the previous run.  Without this the
set was always empty on start, causing every connected minion to appear
"new" and fire spurious presence/change events.

Adds import salt.utils.cache, seeds old_present from the disk cache at
the top of Maintenance.run(), and writes the cache back at the end of
each handle_presence() call.

Adds a parametrized unit test covering no-change, new-minion, and
lost-minion scenarios, and verifies the on-disk cache is updated
correctly.

Fixes #64505
